mlflow/pytorch.py

```diff
@@ -0,0 +1,133 @@
+    :param kwargs: kwargs to pass to `torch.save` method
+    """
+
+    if not isinstance(pytorch_model, torch.nn.Module):
+        raise TypeError("Argument 'pytorch_model' should be a torch.nn.Module")
+
+    path = os.path.abspath(path)
+    if os.path.exists(path):
+        raise Exception("Path '{}' already exists".format(path))
+    os.makedirs(path)
+    model_path = os.path.join(path, "model.pth")
+
+    # Save pytorch model
+    torch.save(pytorch_model, model_path, **kwargs)
+    model_file = os.path.basename(model_path)
+
+    mlflow_model.add_flavor(FLAVOR_NAME, model_data=model_file, pytorch_version=torch.__version__)
+    pyfunc.add_to_model(mlflow_model, loader_module="mlflow.pytorch",
+                        data=model_file, env=conda_env)
+    mlflow_model.save(os.path.join(path, "MLmodel"))
+
+
+def _load_model(path, **kwargs):
+    mlflow_model_path = os.path.join(path, "MLmodel")
+    if not os.path.exists(mlflow_model_path):
+        raise RuntimeError("MLmodel is not found at '{}'".format(path))
+
+    mlflow_model = Model.load(mlflow_model_path)
+
+    assert FLAVOR_NAME in mlflow_model.flavors, \
+        "Stored model can not be loaded with mlflow.pytorch"
+
+    # This maybe replaced by a warning and then try/except torch.load
+    flavor = mlflow_model.flavors[FLAVOR_NAME]
+    assert torch.__version__ == flavor["pytorch_version"], \
+        "Unfortunately stored model version '{}' does not match "\
+        .format(flavor["pytorch_version"]) + \
+        "installed PyTorch version '{}'".format(torch.__version__)
+
+    path = os.path.abspath(path)
+    path = os.path.join(path, mlflow_model.flavors[FLAVOR_NAME]['model_data'])
+    return torch.load(path, **kwargs)
+
+
+def load_model(path, run_id=None, **kwargs):
+    """
+    Load a PyTorch model from a local file (if run_id is None) or a run.
+    :param path: Local filesystem path or Run-relative artifact path to the model saved by
+        `mlflow.pytorch.log_model`.
+    :param run_id: Run ID. If provided it is combined with path to identify the model.
+    :param kwargs: kwargs to pass to `torch.load` method
+    """
+    if run_id is not None:
+        path = mlflow.tracking._get_model_log_dir(model_name=path, run_id=run_id)
+
+    return _load_model(path, **kwargs)
+
+
+def load_pyfunc(path, **kwargs):
+    """
+    Load the model as PyFunc.
+    :param path: Local filesystem path to the model saved by `mlflow.pytorch.log_model`.
+    :param kwargs: kwargs to pass to `torch.load` method
+    """
+    return _PyTorchWrapper(_load_model(os.path.dirname(path), **kwargs))
+
+
+class _PyTorchWrapper(object):
+    """
+    Wrapper class that creates a predict function such that
+    predict(data: ndarray) -> model's output as numpy.ndarray
+    """
+    def __init__(self, pytorch_model):
+        self.pytorch_model = pytorch_model
+
+    def predict(self, data, device='cpu'):
+        assert isinstance(data, np.ndarray), "Input data should be numpy.ndarray"
```

Comment:
Does this support inputting a `pandas.Dataframe`? For [reference](https://www.mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#module-mlflow.pyfunc).

Reply:
@juntai-zheng no, it does not support `pandas.DataFrame`. 
How could we put an image into a dataframe ?

Reply:
This is a good point, but currently implementors of pyfunc do expect that they can hand `predict` a DataFrame and get a DataFrame out.

We should have a follow-up discussion on extending the `predict` interface to be able to support n-dimensional. Until then, perhaps we could have two pathways, one where data is an np.ndarray, and the other where the input is a pandas DataFrame. In the latter case, we can directly convert it to an ndarray, and we also convert the result to a pandas DataFrame. This does assume the input is 2D for now, which is pretty constraining.

Reply:
@aarondav just wonder whether `predict` functions  for other frameworks (sklearn, h2o, tf etc) do not convert DataFrame to numpy array to execute a processing ?
What are the advantages of using DataFrame vs 2D numpy.array ?

Reply:
Right, good question. The idea of taking a DataFrame is to integrate with other frameworks. For example, you can load a pyfunc into a Spark UDF:
https://github.com/mlflow/mlflow/blob/2b5a21af05defcfa80255c081b5d9f07443f3f64/mlflow/pyfunc/__init__.py#L173-L184

In order for this to work, we use the Pandas Vectorized UDF support in Spark to efficiently convert Spark DataFrames to Pandas DataFrames, and then pass them into the predict() function. If the predict() function cannot take Pandas DataFrames, then it simply cannot be used as a Spark UDF.